src/System.Sorting/src/Sort.cs

```diff
@@ -0,0 +1,223 @@
+        /// elements to each other using the IComparable interface, which
+        /// must be implemented by all elements in the given section of the array.
+        /// </summary>
+        /// <param name="array"></param>
+        /// <param name="index"></param>
+        /// <param name="length"></param>
+        public virtual void Sort<T>(T[] array, int index, int length) //where T : IComparable
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+            this.Sort<T>(array, index, length, null);
+        }
+
+              
+        /// <summary>
+        /// Sorts the elements in a section of an array. The sort compares the
+        /// elements to each other using the given IComparer interface. If
+        /// comparer is null, the elements are compared to each other using
+        /// the IComparable interface, which in that case must be implemented
+        /// by all elements in the given section of the array.
+        /// </summary>
+        /// <param name="array"></param>
+        /// <param name="index"></param>
+        /// <param name="length"></param>
+        /// <param name="comparer"></param>
+        public void Sort<T> (T[] array, int index, int length, IComparer<T> comparer) //where T : IComparable
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+            if (index < 0 || length < 0)
+                //throw new ArgumentOutOfRangeException((length < 0 ? "length" : "index"), Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+                throw new ArgumentOutOfRangeException((length < 0 ? "length" : "index"), "ArgumentOutOfRange_NeedNonNegNum");
+            if (array.Length - index < length)
+                //throw new ArgumentException(Environment.GetResourceString("Argument_InvalidOffLen"));
+                throw new ArgumentException("Argument_InvalidOffLen");
+
+            if (length < 2)
+                return;
+            
+            //Delegating the implementation to the derived classes which will implement different types of sorting algorithms. 
+            SortInternal<T>(array, index, length, comparer);
+            
+        }
+
+
+        /// <summary>
+        /// Sorts the elements in a section of an array. The sort compares the
+        /// elements to each other using the given IComparer interface. If
+        /// comparer is null, the elements are compared to each other using
+        /// the IComparable interface, which in that case must be implemented
+        /// by all elements in the given section of the array.
+        /// </summary>
+        /// <param name="array"></param>
+        /// <param name="index"></param>
+        /// <param name="length"></param>
+        /// <param name="comparer"></param>
+        internal virtual void SortInternal<T>(T[] array, int index, int length, IComparer<T> comparer) //where T : IComparable
```

Comment:
It would be better if this was `protected` and `abstract`.  It being `abstract` will require it to be implemented.

The convention we follow for methods were you have to implement them but not do argument checking is to suffix the method with the word "Core", so I would also rename this to SortCore.